src/charm.py

```diff
@@ -356,6 +356,31 @@ def _block_on_openstack_config(self, state: CharmState) -> bool:
             return True
         return False
 
+    def _common_install_code(self, state: CharmState) -> bool:
+        """Installation code shared between install and upgrade hook.
+
+        Returns:
+            Whether the installation code was successful.
+        """
+        self.unit.status = MaintenanceStatus("Installing packages")
+        try:
+            # The `_start_services`, `_install_deps` includes retry.
+            self._install_deps()
+            self._start_services(state.charm_config.token, state.proxy_config)
+            metrics.setup_logrotate()
+        except (LogrotateSetupError, SubprocessError) as err:
+            logger.exception(err)
+            if isinstance(err, LogrotateSetupError):
+                msg = "Failed to setup logrotate"
+            else:
+                msg = "Failed to install dependencies"
+            self.unit.status = BlockedStatus(msg)
```

Comment:
What would be the action to unblock the charm?

Reply:
Thanks, I have updated the code to go into error state.